rskj-core/src/test/java/co/rsk/peg/feeperkb/FeePerKbIntegrationTest.java

```diff
@@ -0,0 +1,478 @@
+package co.rsk.peg.feeperkb;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import co.rsk.bitcoinj.core.Coin;
+import co.rsk.core.RskAddress;
+import co.rsk.peg.InMemoryStorage;
+import co.rsk.peg.feeperkb.constants.FeePerKbConstants;
+import co.rsk.peg.feeperkb.constants.FeePerKbMainNetConstants;
+import co.rsk.peg.storage.StorageAccessor;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.ethereum.core.SignatureCache;
+import org.ethereum.core.Transaction;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class FeePerKbIntegrationTest {
+    private final FeePerKbConstants feePerKbConstants = FeePerKbMainNetConstants.getInstance();
+    private FeePerKbSupport feePerKbSupport;
+    private SignatureCache signatureCache;
+
+    @BeforeEach
+    void setUp() {
+        StorageAccessor inMemoryStorage = new InMemoryStorage();
+        FeePerKbStorageProvider feePerKbStorageProvider = new FeePerKbStorageProviderImpl(inMemoryStorage);
+        feePerKbSupport = new FeePerKbSupportImpl(feePerKbConstants, feePerKbStorageProvider);
+        signatureCache = mock(SignatureCache.class);
+    }
+
+    @Test
+    void feePerKbIntegrationTest() {
+        Coin differentFeePerKbVote = Coin.valueOf(60_000L);
+        Coin maxFeePerKbVote = feePerKbConstants.getMaxFeePerKb();
+        Coin excessiveFeePerKbVote = maxFeePerKbVote.add(Coin.SATOSHI);
+        Coin genesisFeePerKb = feePerKbConstants.getGenesisFeePerKb();
+
+        // Get fee per kb before any voting, should return the genesis fee per kb
+        assertFeePerKbValue(genesisFeePerKb);
+
+        /*
+         *  1st voting: authorizer 1 and 2 vote the same value
```

Comment:
I am wandering, can each of these sections be placed in its own test? Or is there a need for all of them to be chained in this manner?

Reply:
Hey, Antonio!
Although we're working with UT's tools, this approach creates integration tests, that is, test the whole Fee Per KB module, for this reason, it is important to make different test cases stressing the InMemoryStorage executing one test after another, and observe the behavior interacting like a real-life scenario without isolate them.

Reply:
We should this with the rest of the team @apancorb. The plan here was to create a test that simulates a real interaction with the fee per kb feature. That implies keeping the state and not reseting it between tests.

One option is to do it this way, in a single big test executing one action after the other. The other option is to do it in separate tests, sharing the same environment between tests. This last option implies that those tests will not be independent.

What do you think? Any other ideas?

Reply:
I see, for integration tests sharing state between tests is fine. I would recommend having each Xth voting section in its own small test and sharing state at the class level.

Check [this article](https://www.baeldung.com/junit-testinstance-annotation#:~:text=Sharing%20state%20is%20usually%20an,tests%20that%20intentionally%20share%20state.) out that will describe how to set per-class lifecycle and how to set the ordering of tests if it matters. This might help you achieve what you want to create. 

_As a side note, integration tests are usually placed under the `intTest` directory next to `main` and `test`. Gradle natively supports this convention. Maybe in the future this is something we can adopt._

Reply:
I agree with @apancorb. This long test should be split using one of the options available for this case using `junit`.

Reply:
### Refactoring Test Integration for Fee Per KB
- Adding some annotations such as @testinstance, and @BeforeAll to help share state in the test.
- Adding annotations such as @TestMethodOrder, and @order to create an order of test execution.
- Adding Caller enum to get the code more expressive.
- After adding the Caller enum, adapt the voteFeePerKb method to receive as an In Parameter RskAddress.
- Removing voteFeePerKbByUnauthorizedCaller, getAuthorizedRskAddresses, and getUnauthorizedRskAddress methods since they are no longer needed after the adaptation of the Caller enum in the code.

_P.S.: it's just missing to move this integration test to the IntegrationTest directory._